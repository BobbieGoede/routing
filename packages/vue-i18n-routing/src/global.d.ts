declare let __DEV__: boolean
declare let __TEST__: boolean
declare let __VERSION__: string
