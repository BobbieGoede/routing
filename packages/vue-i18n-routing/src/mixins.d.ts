export * from 'vue-i18n-routing'
import 'vue-i18n-routing.mixins'
