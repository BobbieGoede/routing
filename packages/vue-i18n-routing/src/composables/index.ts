export * from './routing'
export * from './routing2'
export { getRouteBaseName } from './utils'
