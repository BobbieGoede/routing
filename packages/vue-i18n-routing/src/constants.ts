export const STRATEGIES = {
  PREFIX: 'prefix',
  PREFIX_EXCEPT_DEFAULT: 'prefix_except_default',
  PREFIX_AND_DEFAULT: 'prefix_and_default',
  NO_PREFIX: 'no_prefix'
} as const

export const VUE_I18N_ROUTING_DEFAULTS = {
  defaultLocale: '',
  strategy: 'prefix_except_default',
  trailingSlash: false,
  routesNameSeparator: '___',
  defaultLocaleRouteNameSuffix: 'default'
}
