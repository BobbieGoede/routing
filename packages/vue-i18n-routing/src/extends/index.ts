export * from './router'
export type { VueI18nRoutingPluginOptions } from './i18n'
