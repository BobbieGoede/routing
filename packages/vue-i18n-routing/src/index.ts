/**
 * Vue I18n Routing Version
 *
 * @remarks
 * Semver format. Same format as the package.json `version` field.
 */
export const VERSION = __VERSION__

export * from './composables'
export * from './resolve'
export * from './types'
export { getLocale, setLocale } from './utils'
export { createRouter, createLocaleFromRouteGetter } from './extends'
