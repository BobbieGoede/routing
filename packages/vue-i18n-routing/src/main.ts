export const VueI18nRouting = {
  routes: []
}

console.log('this is vue-i18n-rouging!')
