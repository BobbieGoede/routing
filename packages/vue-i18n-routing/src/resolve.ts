import { adjustRoutePathForTrailingSlash } from './utils'
import { VUE_I18N_ROUTING_DEFAULTS } from './constants'

import type { VueI18nRoute, VueI18nRoutingOptions } from './types'

// type RouteOptions = {
//   locales: string[]
//   paths: Record<string, any> // eslint-disable-line @typescript-eslint/no-explicit-any
// }

export function localizeRoutes(
  routes: VueI18nRoute[],
  {
    defaultLocale = VUE_I18N_ROUTING_DEFAULTS.defaultLocale,
    trailingSlash = VUE_I18N_ROUTING_DEFAULTS.trailingSlash,
    routesNameSeparator = VUE_I18N_ROUTING_DEFAULTS.routesNameSeparator,
    defaultLocaleRouteNameSuffix = VUE_I18N_ROUTING_DEFAULTS.defaultLocaleRouteNameSuffix,
    localeCodes = []
  }: Pick<
    VueI18nRoutingOptions,
    'defaultLocale' | 'localeCodes' | 'routesNameSeparator' | 'trailingSlash' | 'defaultLocaleRouteNameSuffix'
  > = {}
): VueI18nRoute[] {
  function makeLocalizedRoutes(
    route: VueI18nRoute,
    allowedLocaleCodes: string[],
    isChild = false,
    isExtraPageTree = false
  ): VueI18nRoute[] {
    // skip route localization
    if (route.redirect && (!route.component || !route.file)) {
      return [route]
    }

    // TODO: route from options

    const targetLocales = allowedLocaleCodes

    // TODO: component options

    return targetLocales.reduce((_routes, locale) => {
      const { name } = route
      let { path } = route
      const localizedRoute = { ...route }

      // make localized page name
      if (name) {
        localizedRoute.name = `${name}${routesNameSeparator}${locale}`
      }

      // generate localized children routes
      if (route.children) {
        localizedRoute.children = route.children.reduce(
          (children, child) => [...children, ...makeLocalizedRoutes(child, [locale], true, isExtraPageTree)],
          [] as NonNullable<VueI18nRoute['children']>
        )
      }

      // TODO: custom paths

      const isDefaultLocale = locale === defaultLocale
      if (isDefaultLocale) {
        if (!isChild) {
          const defaultRoute = { ...localizedRoute, path }

          if (name) {
            defaultRoute.name = `${localizedRoute.name}${routesNameSeparator}${defaultLocaleRouteNameSuffix}`
          }

          if (route.children) {
            // recreate child routes with default suffix added
            defaultRoute.children = []
            for (const childRoute of route.children) {
              defaultRoute.children = defaultRoute.children.concat(
                makeLocalizedRoutes(childRoute as VueI18nRoute, [locale], true, true)
              )
            }
          }

          _routes.push(defaultRoute)
        } else if (isChild && isExtraPageTree && name) {
          localizedRoute.name += `${routesNameSeparator}${defaultLocaleRouteNameSuffix}`
        }
      }

      const isChildWithRelativePath = isChild && !path.startsWith('/')

      // add route prefix
      const shouldAddPrefix = !isChildWithRelativePath && !isDefaultLocale
      if (shouldAddPrefix) {
        path = `/${locale}${path}`
      }

      if (path) {
        path = adjustRoutePathForTrailingSlash(path, trailingSlash, isChildWithRelativePath)
      }

      localizedRoute.path = path
      _routes.push(localizedRoute)

      return _routes
    }, [] as VueI18nRoute[])
  }

  return routes.reduce(
    (localized, route) => [...localized, ...makeLocalizedRoutes(route, localeCodes || [])],
    [] as VueI18nRoute[]
  )
}
