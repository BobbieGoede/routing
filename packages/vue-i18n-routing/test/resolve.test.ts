/// <reference path="../../../test/chai.shim.d.ts"/>

import { expect } from 'chai'
import { localizeRoutes } from '../src/resolve'
import { VUE_I18N_ROUTING_DEFAULTS } from '../src/constants'

import type { VueI18nRoute } from '../src/types'

describe('localizeRoutes', function () {
  describe('basic', function () {
    it('should be localized routing', function () {
      const routes: VueI18nRoute[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, { localeCodes })

      expect(localizedRoutes).to.matchSnapshot(this)
      expect(localizedRoutes.length).to.equal(4)
      localeCodes.forEach(locale => {
        routes.forEach(route => {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}`,
            name: `${route.name}${VUE_I18N_ROUTING_DEFAULTS.routesNameSeparator}${locale}`
          })
        })
      })
    })
  })

  describe('has children', function () {
    it('should be localized routing', function () {
      const routes: VueI18nRoute[] = [
        {
          path: '/user/:id',
          name: 'user',
          children: [
            {
              path: 'profile',
              name: 'user-profile'
            },
            {
              path: 'posts',
              name: 'user-posts'
            }
          ]
        }
      ]
      const children: VueI18nRoute[] = routes[0].children as VueI18nRoute[]

      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, { localeCodes })

      expect(localizedRoutes).to.matchSnapshot(this)
      expect(localizedRoutes.length).to.equal(2)
      localeCodes.forEach(locale => {
        routes.forEach(route => {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}`,
            name: `${route.name}${VUE_I18N_ROUTING_DEFAULTS.routesNameSeparator}${locale}`,
            children: children.map(child => ({
              path: child.path,
              name: `${child.name}${VUE_I18N_ROUTING_DEFAULTS.routesNameSeparator}${locale}`
            }))
          })
        })
      })
    })
  })

  describe('trailing slash', function () {
    it('should be localized routing', function () {
      const routes: VueI18nRoute[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, { localeCodes, trailingSlash: true })

      expect(localizedRoutes).to.matchSnapshot(this)
      expect(localizedRoutes.length).to.equal(4)
      localeCodes.forEach(locale => {
        routes.forEach(route => {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}/`,
            name: `${route.name}${VUE_I18N_ROUTING_DEFAULTS.routesNameSeparator}${locale}`
          })
        })
      })
    })
  })

  describe('route name separator', function () {
    it('should be localized routing', function () {
      const routes: VueI18nRoute[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, { localeCodes, routesNameSeparator: '__' })

      expect(localizedRoutes).to.matchSnapshot(this)
      expect(localizedRoutes.length).to.equal(4)
      localeCodes.forEach(locale => {
        routes.forEach(route => {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}`,
            name: `${route.name}${'__'}${locale}`
          })
        })
      })
    })
  })

  describe('strategy: "prefix_and_default"', function () {
    it('should be localized routing', function () {
      const routes: VueI18nRoute[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        defaultLocale: 'en',
        strategy: 'prefix_and_default',
        localeCodes
      })

      expect(localizedRoutes).to.matchSnapshot(this)
    })
  })

  describe('strategy: "prefix"', function () {
    it('should be localized routing', function () {
      const routes: VueI18nRoute[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        defaultLocale: 'en',
        strategy: 'prefix',
        localeCodes,
        includeUprefixedFallback: true
      })

      expect(localizedRoutes).to.matchSnapshot(this)
    })
  })

  describe('strategy: "no_prefix"', function () {
    it('should be localized routing', function () {
      const routes: VueI18nRoute[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        defaultLocale: 'en',
        strategy: 'no_prefix',
        localeCodes
      })

      expect(localizedRoutes).to.matchSnapshot(this)
    })
  })
})
