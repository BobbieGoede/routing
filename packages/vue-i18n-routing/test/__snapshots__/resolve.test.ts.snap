exports[`localizeRoutes : basic : should be localized routing 1`] = `[
  {
    "path": "/en",
    "name": "home___en"
  },
  {
    "path": "/ja",
    "name": "home___ja"
  },
  {
    "path": "/en/about",
    "name": "about___en"
  },
  {
    "path": "/ja/about",
    "name": "about___ja"
  }
]`;

exports[`localizeRoutes : has children : should be localized routing 1`] = `[
  {
    "path": "/en/user/:id",
    "name": "user___en",
    "children": [
      {
        "path": "profile",
        "name": "user-profile___en"
      },
      {
        "path": "posts",
        "name": "user-posts___en"
      }
    ]
  },
  {
    "path": "/ja/user/:id",
    "name": "user___ja",
    "children": [
      {
        "path": "profile",
        "name": "user-profile___ja"
      },
      {
        "path": "posts",
        "name": "user-posts___ja"
      }
    ]
  }
]`;

exports[`localizeRoutes : trailing slash : should be localized routing 1`] = `[
  {
    "path": "/en/",
    "name": "home___en"
  },
  {
    "path": "/ja/",
    "name": "home___ja"
  },
  {
    "path": "/en/about/",
    "name": "about___en"
  },
  {
    "path": "/ja/about/",
    "name": "about___ja"
  }
]`;

exports[`localizeRoutes : route name separator : should be localized routing 1`] = `[
  {
    "path": "/en",
    "name": "home__en"
  },
  {
    "path": "/ja",
    "name": "home__ja"
  },
  {
    "path": "/en/about",
    "name": "about__en"
  },
  {
    "path": "/ja/about",
    "name": "about__ja"
  }
]`;

exports[`localizeRoutes : strategy: "prefix_and_default" : should be localized routing 1`] = `[
  {
    "path": "/",
    "name": "home___en___default"
  },
  {
    "path": "/en",
    "name": "home___en"
  },
  {
    "path": "/ja",
    "name": "home___ja"
  },
  {
    "path": "/about",
    "name": "about___en___default"
  },
  {
    "path": "/en/about",
    "name": "about___en"
  },
  {
    "path": "/ja/about",
    "name": "about___ja"
  }
]`;

exports[`localizeRoutes : strategy: "prefix" : should be localized routing 1`] = `[
  {
    "path": "/",
    "name": "home"
  },
  {
    "path": "/en",
    "name": "home___en"
  },
  {
    "path": "/ja",
    "name": "home___ja"
  },
  {
    "path": "/about",
    "name": "about"
  },
  {
    "path": "/en/about",
    "name": "about___en"
  },
  {
    "path": "/ja/about",
    "name": "about___ja"
  }
]`;

exports[`localizeRoutes : strategy: "no_prefix" : should be localized routing 1`] = `[
  {
    "path": "/",
    "name": "home"
  },
  {
    "path": "/about",
    "name": "about"
  }
]`;

